### PR TITLE
Port chromium message parser and refactor

### PR DESCRIPTION
**Pull request by ibash**
This commit:
1. Adds sanity tests against chrome gcm. To use the tests copy
   test/keys.template.js to test/keys.js and populate it with gcm
   credentials / sender id. Then do `npm run test`

2. Ports chromium's ConnectionHandlerImpl::WaitForData. This fixes the
   problem of not parsing the wire format correctly. There are two
   more things here that would be nice to port:
   - Add timeouts while waiting for message bytes
   - Error handling around protobufs

   I did make the simplification of not supporting the
   kDefaultDataPacketLimit limit (it shouldn't affect correctness)

   ref: https://cs.chromium.org/chromium/src/google_apis/gcm/engine/connection_handler_impl.cc?rcl=dc7c41bc0ee5fee0ed269495dde6b8c40df43e40&l=178

3. Refactors client and socket into a class. Combining them made it
   easier to cleanup the socket at the end of the test. It also has the
   benefit of making the retrying a bit cleaner.

For the code reviewer: I'd love to have a second pair of eyes comparing the port to the original from https://cs.chromium.org/chromium/src/google_apis/gcm/engine/connection_handler_impl.cc?rcl=dc7c41bc0ee5fee0ed269495dde6b8c40df43e40&l=178. In particular - is there anything major I missed? Is there anything slightly different? Any off by one errors?

### CODE REVIEW

File: src/client.js [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+      // FIXME Figure out how to pass persistentIds without being kickout by Google
+      receivedPersistentId : [],
+    };
+
+    const errorMessage = LoginRequestType.verify(loginRequest);
+    if (errorMessage) {
+      throw new Error(errorMessage);
+    }
+    const message = LoginRequestType.create(loginRequest);
+    const buffer = LoginRequestType.encode(message).finish();
```
Comment (yurynix) [inline src/client.js]: You can `LoginRequestType.encodeDelimited(loginRequest).finish()` that will include the varint size, and drop the `149, 1` from the following buffer

Comment (ibash) [reply src/client.js]: Ah good call - will make that change, and will use `kMCSVersion` and `kLoginRequestTag` from parser

File: src/index.js [hunk 2]
```diff
@@ -30,7 +28,9 @@ async function listen(credentials, notificationCallback) {
...
     throw new Error('Missing keys.authSecret in credentials');
   }
-  await checkIn(credentials.gcm.androidId, credentials.gcm.securityToken);
-  emitter.on(ON_NOTIFICATION_RECEIVED, notificationCallback);
-  await connect(credentials.gcm, credentials.keys, credentials.persistentIds);
+
+  const client = new Client(credentials, credentials.persistentIds);
+  client.on('ON_NOTIFICATION_RECEIVED', notificationCallback);
+  client.connect();
+  return client;
```
Comment (ibash) [inline src/index.js]: Since client is returned here you can get #7 for free by calling client.destroy()

**Comment by MatthieuLemoine** [comment]
Thanks @ibash ! I will look at it this weekend

File: src/parser.js [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+  kHeartbeatAckTag,
+  kLoginRequestTag,
+  kLoginResponseTag,
+  kCloseTag,
+  kIqStanzaTag,
+  kDataMessageStanzaTag,
+  kStreamErrorStanzaTag,
+} = require('./constants');
+
+const DEBUG = () => {};
```
Comment (MatthieuLemoine) [inline src/parser.js]: We could use https://github.com/visionmedia/debug instead

**Comment by MatthieuLemoine** [comment]
@ibash Nice work ! Also thanks for the tests. `push-receiver` is way more reliable now 👍 I just made a few format changes using prettier & replace mocha with jest. I also added a precommit hook to run eslint & jest.

Could you update the documentation too ? Now that `client` is returned by `listen` we can subscribe to some events like 'connect' & we can call destroy to stop listening.

File: src/utils/decrypt/index.js [hunk 4]
```diff
@@ -5,32 +5,21 @@ module.exports = decrypt;
...
-      console.error('Error while decrypting incoming message');
-      console.error(e);
-      return;
-    }
-  } catch (e) {
-    // Not a notification
-    return;
-  }
+  const cryptoKey = object.appData
+    .find(item => item.key === 'crypto-key')
```
Comment (yurynix) [inline src/utils/decrypt/index.js]: `item.key === 'crypto-key'` is not necessarily present, at least on GCM pushes.

File: src/utils/decrypt/index.js [hunk 5]
```diff
@@ -5,32 +5,21 @@ module.exports = decrypt;
...
+    .value.slice(3);
+  const salt = object.appData
+    .find(item => item.key === 'encryption')
```
Comment (yurynix) [inline src/utils/decrypt/index.js]: Same here, it's not necessarily present on GCM pushes...

**Comment by yurynix** [comment]
When I got pushes from GCM it wasn't encrypted at all, so, decrypt was failing as well for me, I just removed it.

So, we should check if push contains `encryption` and `crypto-key` and not assume it has them.

Here's an example GCM DataMessageStanza 
```json
{
  "appData": [
    {
      "key": "expires",
      "value": "1518341280"
    },
    {
      "key": "<some app specific key>",
      "value": "<some app specific data>"
    }
  ],
  "category": "com.company.android",
  "from": "<sender id>",
  "fromTrustedServer": false,
  "id": "850A672B",
  "persistentId": "0:1518341028984788%8d385070f9fd7ecd",
  "sent": "1518341028979",
  "ttl": 2419200
}
```

**Comment by MatthieuLemoine** [comment]
@ibash could you check if the messages you receive that throw contains a cryptoKey? If not we could add a check

**Comment by MatthieuLemoine** [comment]
Also see #8

**Comment by ibash** [comment]
@MatthieuLemoine / @yurynix I'm adding some logging to check if that's the case for when I'm seeing the error, will let you know when I get these errors in

**Comment by ibash** [comment]
So - the crypto-key / salt definitely exist on the notifications that are failing to decrypt. I spent some time digging through how chromium encrypts/decrypts gcm messages. I didn't see anything obvious, I think my next step is to log whenever the keys change just to make sure it's not something like the notification being encrypted for older keys.

**Comment by ibash** [comment]
Putting this on the backburner on our side as it's not a hard requirement, I'll probably pick this back up later this week or mid next week

**Comment by MatthieuLemoine** [comment]
@ibash any news ?

**Comment by ibash** [comment]
@MatthieuLemoine I've been digging into this over the last week or so.
So far I've confirmed that:
1. The keys are fine, I'm able to receive other notifications using the same keys.
2. Chrome's own crypto code can't decrypt the notification with the keys  / cipher text (I made a unit test in chrome's code and hardcoded values that should work, and then tried the values that weren't working for me). This implies that the http_ece code is correct, too.
3. I tried decrypting the notification that fails with other keys used on my machine, and it also fails, so I don't think it's a problem with the notification being for the wrong key.
4. I did some digging through chrome's code and it looks like if they fail to decrypt a notification they just log it and drop the notifications.
5. I added some more logging to see if after getting a notification that fails to decrypt, all future notifications fail, or if the code can recover from this error. So far it looks like the code can recover, but I'm waiting to get more data to see how often decrypting fails after that.

This makes me think that it's a problem with either:
1. The ciphertext is not being parsed corrrectly (and so it fails when trying to decrypt).
2. Chrome's server is sending something that we can't decrypt.

Any thoughts / ideas?

**Comment by MatthieuLemoine** [comment]
@ibash Are you sure that you're missing a notification ? Maybe this message is not a valid notification.

**Comment by ibash** [comment]
@MatthieuLemoine I'm fairly sure - they have a category that we set. In anycase I'm deciding to silently drop it, since we still get most notifications.

**Comment by ibash** [comment]
@MatthieuLemoine / @yurynix updated the PR, WDYT?

**Comment by ibash** [comment]
I started clearing persistent ids on our fork: https://github.com/superhuman/push-receiver/pull/6/files

**Comment by yurynix** [comment]
Regarding the decryption issue, 
What's the step to reproduce it?

Just run this branch and send notifications from FCM? After how many I should get an error?

File: src/client.js [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    };
+
+    const errorMessage = LoginRequestType.verify(loginRequest);
+    if (errorMessage) {
+      throw new Error(errorMessage);
+    }
+    const message = LoginRequestType.create(loginRequest);
+    const buffer = LoginRequestType.encodeDelimited(message).finish();
+
+    // FIXME Can change depending on persistentIds
```
Comment (yurynix) [inline src/client.js]: This comment no longer relevant IMO,
`encodeDelimited` handles the varint size, so the code here handles it ok

File: src/client.js [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+      // FIXME Figure out how to pass persistentIds without being kickout by Google
+      receivedPersistentId : [],
+    };
+
+    const errorMessage = LoginRequestType.verify(loginRequest);
+    if (errorMessage) {
+      throw new Error(errorMessage);
+    }
+    const message = LoginRequestType.create(loginRequest);
+    const buffer = LoginRequestType.encodeDelimited(message).finish();
```
Comment (yurynix) [inline src/client.js]: its ok to have here:
`const buffer = LoginRequestType.encodeDelimited(loginRequest).finish();`
  no need to
` const message = LoginRequestType.create(loginRequest);`

File: src/client.js [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+      domain               : 'mcs.android.com',
+      deviceId             : `android-${hexAndroidId}`,
+      networkType          : 1,
+      resource             : this._credentials.gcm.androidId,
+      user                 : this._credentials.gcm.androidId,
+      useRmq2              : true,
+      setting              : [{ name : 'new_vc', value : '1' }],
+      // Id of the last notification received
+      clientEvent          : [],
+      // FIXME Figure out how to pass persistentIds without being kickout by Google
```
Comment (yurynix) [inline src/client.js]: that comment no longer relevant, the code already handles that, just need to pass it

**Comment by MatthieuLemoine** [comment]
@ibash Can you add the persistentIds clearing in this PR ?

**Comment by MatthieuLemoine** [comment]
@ibash Is this PR ready to be merged ?

**Comment by ibash** [comment]
@MatthieuLemoine I think your test changes made them more leaky, as they now reach into global scope. Going to change them back (as it's making the tests less reliable for me...)

**Comment by ibash** [comment]
@yurynix I haven't found a way to reproduce the encryption issue reliably, it seems random AFAICT (maybe upon initial connection? I saw it once in dev mode right after booting the app).

**Comment by ibash** [comment]
@MatthieuLemoine Ah - didn't realize jest doesn't have a context for each test, going to make it just clear credentials.

**Comment by ibash** [comment]
I think so, thank you!

Islam Sharabash
217-377-9657

Sent via Superhuman iOS ( https://sprh.mn/?vip=islam.sharabash@gmail.com )

**Pull request merged**